### PR TITLE
Elasticsearch index must be lowercase

### LINKED ISSUE #6342: Automatically lowercase the index name when custom fields is used to build it
As uppercase is invalid in elasticsearch indices, Beats should either provide a lowercase processor or always ensure index names are lowercased when using custom fields

Use case (filebeat):

A document:
```
{"HDG": 125.0, "fwot": "FOO", "date": "2018-01-31 23:00:03"}
```

A beat configuration:

```
.......
- type: log
  paths:
    - /var/log/myfeed/*.log
  json.keys_under_root: true
......
output.elasticsearch:
  hosts: ["localhost:9200"]
  protocol: "http"
  username: "elastic"
  password: "changeme"
  index: "custom-%{[fwot]}-%{+yyyy.MM}-v1"
.....
```

### PR DESCRIPTION
**Pull request by urso**
- Bug

## What does this PR do?

This change ensure that index names are always converted to lowercase.
Static strings are converted to lowercase upfront, while dynamic strings
will be post-processed.

## Why is it important?

When indexing into Elasticsearch index names must always be lowercase.
If the index or indices setting are configured to produce non-lowercase
strings (e.g. by extracting part of the index name from the event
contents), we need to normalize them to be lowercase.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## How to test this PR locally

Index with different index/indices configurations into Elasticsearch:
- use upper case in `index` setting
- extract index name from event, that has the field value as upper case string
- configure a mapping via `indices` where the target index is upper case

Run same checks with Redis, Kafka, Logstash outputs (the index name should consistently be lower case for all outputs).

## Related issues

- Closes elastic/beats#6342 
- Closes elastic/beats#6640

### CODE REVIEW

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -50,6 +50,8 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Update replicaset group to apps/v1 {pull}15854[15802]
 - Fix issue where default go logger is not discarded when either * or stdout is selected. {issue}10251[10251] {pull}15708[15708]
 - Upgrade go-ucfg to latest v0.8.1. {pull}15937{15937}
+- Remove superfluous use of number_of_routing_shards setting from the default template. {pull}16038[16038]
```
Comment (kvch) [inline CHANGELOG.next.asciidoc]: This changelog entry does not belong to the PR.

Comment (urso) [reply CHANGELOG.next.asciidoc]: Yeah, I pushed some wrong commits to my master by accident. About to clean this up :)

Comment (urso) [reply CHANGELOG.next.asciidoc]: Done.

File: libbeat/outputs/outil/select.go [hunk 2]
```diff
@@ -274,49 +303,27 @@ func buildSingle(cfg *common.Config, key string) (SelectorExpr, error) {
...
-			if err != nil {
-				return nil, err
-			}
-
-			str = mapping.Table[str]
-			if str == "" {
-				str = otherwise
-			}
-
-			if str == "" {
```
Comment (kvch) [inline libbeat/outputs/outil/select.go]: I might be too early to click on approve. Don't we need to set `sel` to `nilSelector` if `otherwise` is an empty string?

Comment (urso) [reply libbeat/outputs/outil/select.go]: We have had this pattern in multiple locations. I moved some of the logic into the expression constructors. e.g. `constSelectorExpr` will check if `str` is "" and create a `nilSelector` or `constSelector`. This change makes code (hopefully) a little easier to read, while reducing duplication and chance of errors.

**Comment by urso** [comment]
Failing tests due to fixes in the ES timestamp parsing. These will be fixed by: https://github.com/elastic/beats/pull/16139
